gem/optimise.py

```diff
@@ -260,6 +286,76 @@ def count(pair):
     return product, flops
 
 
+def associate_sum(summands):
+    """Apply associativity rules to construct an operation-minimal summation tree.
+    """
+    # group summands by their free indices
+    groups = OrderedDict()
+    for summand in summands:
+        key = frozenset(summand.free_indices)
+        groups.setdefault(key, []).append(summand)
+    summands = [reduce(Sum, _value, Zero()) for _value in itervalues(groups)]
+
+    if len(summands) > 32:
+        # O(N^3) algorithm
+        raise NotImplementedError("Not expected such a complicated expression!")
+
+    def count(pair):
+        """Operation count to add a pair of GEM expressions"""
+        a, b = pair
+        extents = [i.extent for i in set().union(a.free_indices, b.free_indices)]
+        return numpy.prod(extents, dtype=int)
+
+    flops = 0
+    while len(summands) > 1:
+        # Greedy algorithm: choose a pair of factors that are the
+        # cheapest to add.
+        a, b = min(combinations(summands, 2), key=count)
+        flops += count((a, b))
+        # Remove chosen factors, append their product
+        summands.remove(a)
+        summands.remove(b)
+        summands.append(Sum(a, b))
+    sum, = summands
+    return sum, flops
+
+
+@singledispatch
+def _reassociate_product(node, self):
+    raise AssertionError("cannot handle type %s" % type(node))
+
+
+_reassociate_product.register(Node)(reuse_if_untouched)
+
+
+@_reassociate_product.register(Product)
+def _reassociate_product_prod(node, self):
+    # collect all factors of product
+    _, factors = traverse_product(node, self.stop_at)
+    factors = list(map(self, factors))
+    product, _ = associate_product(factors)
+    return product
+
+
+def reassociate_product(expressions):
+    """Refactor a Product expression recursively by calling
+    ``associate_product()``
+    :param node: list of expressions
+    :return: list of reassociated product nodes
+    """
+    refcount = collect_refcount(expressions)
+
+    def stop_at(node):
+        if not isinstance(node, Product):
+            return True
+        if refcount[node] > 1:
+            return True
+        return False
+    mapper = Memoizer(_reassociate_product)
+    mapper.stop_at = stop_at
```

Comment:
`stop_at` does not depend on the function arguments of this function, so no need to attach this onto the mapper as an attribute, you could refer to this function directly at use.